kv/kvserver: TestConsistencyQueueRecomputeStats failed

kv/kvserver.TestConsistencyQueueRecomputeStats [failed](https://tanzanite.cluster.engflow.com/invocations/default/19c59fd8-74b2-488c-98d1-c5bf93a4d383?testReportRun=10&testReportShard=47&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [b3fec61ca90095c664f6432af864f18e9946f8bb](https://github.com/cockroachdb/cockroach/commits/b3fec61ca90095c664f6432af864f18e9946f8bb):


```
=== RUN   TestConsistencyQueueRecomputeStats
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestConsistencyQueueRecomputeStats987288777
    test_log_scope.go:76: use -show-logs to present logs inline
    consistency_queue_test.go:445: -- test log scope end --
test logs left over in: outputs.zip/logTestConsistencyQueueRecomputeStats987288777
--- FAIL: TestConsistencyQueueRecomputeStats (31.30s)
=== RUN   TestConsistencyQueueRecomputeStats/hadEstimates=true
    consistency_queue_test.go:619: still have a SysCount of 123010
    --- FAIL: TestConsistencyQueueRecomputeStats/hadEstimates=true (13.14s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=47</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConsistencyQueueRecomputeStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45155

kv/kvserver.TestConsistencyQueueRecomputeStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18028401?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18028401?buildTab=artifacts#/) on master @ [d87d47386f47e826d4694fa50113a1d9ff6a9612](https://github.com/cockroachdb/cockroach/commits/d87d47386f47e826d4694fa50113a1d9ff6a9612):


```
=== RUN   TestConsistencyQueueRecomputeStats
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/1a39945979bd2ca5a0139df86f748a97/logTestConsistencyQueueRecomputeStats2347586378
    test_log_scope.go:76: use -show-logs to present logs inline
    consistency_queue_test.go:445: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1a39945979bd2ca5a0139df86f748a97/logTestConsistencyQueueRecomputeStats2347586378
--- FAIL: TestConsistencyQueueRecomputeStats (27.90s)
=== RUN   TestConsistencyQueueRecomputeStats/hadEstimates=true
    consistency_queue_test.go:619: still have a SysCount of 123010
    --- FAIL: TestConsistencyQueueRecomputeStats/hadEstimates=true (13.68s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConsistencyQueueRecomputeStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestConsistencyQueueRecomputeStats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18068562?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18068562?buildTab=artifacts#/) on master @ [bca6879b87657ac95e6ea756cbb91c9cfa274cfd](https://github.com/cockroachdb/cockroach/commits/bca6879b87657ac95e6ea756cbb91c9cfa274cfd):


```
=== RUN   TestConsistencyQueueRecomputeStats
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/1a39945979bd2ca5a0139df86f748a97/logTestConsistencyQueueRecomputeStats3921052713
    test_log_scope.go:76: use -show-logs to present logs inline
    consistency_queue_test.go:445: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1a39945979bd2ca5a0139df86f748a97/logTestConsistencyQueueRecomputeStats3921052713
--- FAIL: TestConsistencyQueueRecomputeStats (28.71s)
=== RUN   TestConsistencyQueueRecomputeStats/hadEstimates=false
    consistency_queue_test.go:619: still have a SysCount of 123010
    --- FAIL: TestConsistencyQueueRecomputeStats/hadEstimates=false (13.63s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConsistencyQueueRecomputeStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvserver/consistency_queue_test.go
+++ b/pkg/kv/kvserver/consistency_queue_test.go
@@ -606,18 +606,25 @@ func testConsistencyQueueRecomputeStatsImpl(t *testing.T, hadEstimates bool) {
 		t.Fatal(err)
 	}
 
-	// Force a run of the consistency queue, otherwise it might take a while.
-	store := tc.GetFirstStoreFromServer(t, 0)
-	require.NoError(t, store.ForceConsistencyQueueProcess())
+	// When running with leader leases, it might take an extra election interval
+	// for a lease to be established after adding the voters above because the
+	// leader needs to get store liveness support from the followers. The stats
+	// re-computation runs on the leaseholder and will fail if there isn't one.
+	testutils.SucceedsSoon(t, func() error {
+		// Force a run of the consistency queue, otherwise it might take a while.
+		store := tc.GetFirstStoreFromServer(t, 0)
+		require.NoError(t, store.ForceConsistencyQueueProcess())
 
-	// The stats should magically repair themselves. We'll first do a quick check
-	// and then a full recomputation.
-	repl, _, err := tc.Servers[0].GetStores().(*kvserver.Stores).GetReplicaForRangeID(ctx, rangeID)
-	require.NoError(t, err)
-	ms := repl.GetMVCCStats()
-	if ms.SysCount >= sysCountGarbage {
-		t.Fatalf("still have a SysCount of %d", ms.SysCount)
-	}
+		// The stats should magically repair themselves. We'll first do a quick check
+		// and then a full recomputation.
+		repl, _, err := tc.Servers[0].GetStores().(*kvserver.Stores).GetReplicaForRangeID(ctx, rangeID)
+		require.NoError(t, err)
+		ms := repl.GetMVCCStats()
+		if ms.SysCount >= sysCountGarbage {
+			return errors.Newf("still have a SysCount of %d", ms.SysCount)
+		}
+		return nil
+	})
 
 	if delta := computeDelta(db0); delta != (enginepb.MVCCStats{}) {
 		t.Fatalf("stats still in need of adjustment: %+v", delta)

